# Data Mining 201 Projects

# Project 1

## Question 3

In [6]:
import pandas as pd
df = pd.read_csv('/anvil/projects/tdm/data/flights/subset/1991.csv')

In [7]:
df[df["Month"]==12].head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4657380  1991     12          29          7   1631.0        1630   1829.0   
4657381  1991     12          30          1   1630.0        1630   1823.0   
4657382  1991     12          31          2   1628.0        1630   1820.0   
4657383  1991     12           1          7   1922.0        1922   2040.0   
4657384  1991     12           2          1   2005.0        1922   2119.0   

         CRSArrTime UniqueCarrier  FlightNum  ...  TaxiIn  TaxiOut  Cancelled  \
4657380        1828            UA       1770  ...     NaN      NaN          0   
4657381        1828            UA       1770  ...     NaN      NaN          0   
4657382        1828            UA       1770  ...     NaN      NaN          0   
4657383        2045            UA       1774  ...     NaN      NaN          0   
4657384        2045            UA       1774  ...     NaN      NaN          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
4657380               NaN         0           NaN          NaN      NaN   
4657381               NaN         0           NaN          NaN      NaN   
4657382               NaN         0           NaN          NaN      NaN   
4657383               NaN         0           NaN          NaN      NaN   
4657384               NaN         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
4657380            NaN                NaN  
4657381            NaN                NaN  
4657382            NaN                NaN  
4657383            NaN                NaN  
4657384            NaN                NaN  

[5 rows x 29 columns]

4657380	1991	12	29	7	1631.0	1630	1829.0	1828	UA	1770	...	NaN	NaN	0	NaN	0	NaN	NaN	NaN	NaN	NaN
4657381	1991	12	30	1	1630.0	1630	1823.0	1828	UA	1770	...	NaN	NaN	0	NaN	0	NaN	NaN	NaN	NaN	NaN
4657382	1991	12	31	2	1628.0	1630	1820.0	1828	UA	1770	...	NaN	NaN	0	NaN	0	NaN	NaN	NaN	NaN	NaN
4657383	1991	12	1	7	1922.0	1922	2040.0	2045	UA	1774	...	NaN	NaN	0	NaN	0	NaN	NaN	NaN	NaN	NaN
4657384	1991	12	2	1	2005.0	1922	2119.0	2045	UA	1774	...	NaN	NaN	0	NaN	0	NaN	NaN	NaN	NaN	NaN

In [8]:
%sql sqlite:////anvil/projects/tdm/data/movies_and_tv/imdb.db

In [9]:
%%sql

-- get all episodes called "Finale"
SELECT *
FROM episodes AS e
INNER JOIN titles AS t
ON t.title_id = e.episode_title_id
WHERE t.primary_title = 'Finale'
LIMIT 5;

 * sqlite:////anvil/projects/tdm/data/movies_and_tv/imdb.db
Done.


[('tt0506978', 'tt0285332', 1, 22, 'tt0506978', 'tvEpisode', 'Finale', 'Finale', 0, 2002, None, None, 'Drama,Thriller'),
 ('tt0519805', 'tt0210405', 3, 10, 'tt0519805', 'tvEpisode', 'Finale', 'Finale', 0, 1993, None, None, 'Drama'),
 ('tt0566236', 'tt0169438', 8, 31, 'tt0566236', 'tvEpisode', 'Finale', 'Finale', 0, 2005, None, None, 'Drama,Sport'),
 ('tt0566422', 'tt0111945', 9, 26, 'tt0566422', 'tvEpisode', 'Finale', 'Finale', 0, 2004, None, None, 'Comedy,Drama'),
 ('tt0581864', 'tt0044263', 2, 1, 'tt0581864', 'tvEpisode', 'Finale', 'Finale', 0, 1953, None, 30, 'Comedy,Crime,Drama')]

tt0506978	tt0285332	1	22	tt0506978	tvEpisode	Finale	Finale	0	2002	None	None	Drama,Thriller
tt0519805	tt0210405	3	10	tt0519805	tvEpisode	Finale	Finale	0	1993	None	None	Drama
tt0566236	tt0169438	8	31	tt0566236	tvEpisode	Finale	Finale	0	2005	None	None	Drama,Sport
tt0566422	tt0111945	9	26	tt0566422	tvEpisode	Finale	Finale	0	2004	None	None	Comedy,Drama
tt0581864	tt0044263	2	1	tt0581864	tvEpisode	Finale	Finale	0	1953	None	30	Comedy,Crime,Drama

In [10]:
%%bash

names="John Doe;Bill Withers;Arthur Morgan;Mary Jane;Rick Ross;John Marston"
echo $names | cut -d ';' -f 3
echo $names | cut -d ';' -f 6

Arthur Morgan
John Marston


Arthur Morgan
John Marston

## Question 7


In [13]:
import pandas as pd

def summarize_dataframe(dataframe):
    mean_dep_delay = dataframe['DepDelay'].mean()
    median_dep_delay = dataframe['DepDelay'].median()
    std_dev_dep_delay = dataframe['DepDelay'].std()
    shape = dataframe.shape
    
    print("MyDF Summary Statistics ---")
    print(f"Shape:({shape[0]} rows, {shape[1]} columns)")
    print(f"Mean: {mean_dep_delay:.3f}")
    print(f"Median: {median_dep_delay:.3f}")
    print(f"Standard Deviation: {std_dev_dep_delay:.3f}")
    print("---------------------------")

file_path = "/anvil/projects/tdm/data/flights/subset/1991.csv"
df = pd.read_csv(file_path)

summarize_dataframe(df)


MyDF Summary Statistics ---
Shape:(5076925 rows, 29 columns)
Mean: 5.754
Median: 0.000
Standard Deviation: 20.611
---------------------------


In [16]:
%%R

summarize_dataframe <- function(dataframe) {
    mean_dep_delay <- mean(dataframe$DepDelay, na.rm = TRUE)
    median_dep_delay <- median(dataframe$DepDelay, na.rm = TRUE)
    std_dev_dep_delay <- sd(dataframe$DepDelay, na.rm = TRUE)
    shape <- dim(dataframe)
    
    cat("MyDF Summary Statistics ---\n")
    cat("Shape:(", shape[1], "rows,", shape[2], "columns)\n")
    cat("Mean:", mean_dep_delay, "\n")
    cat("Median:", median_dep_delay, "\n")
    cat("Standard Deviation:", std_dev_dep_delay, "\n")
    cat("---------------------------\n")
}

file_path <- "/anvil/projects/tdm/data/flights/subset/1991.csv"
df <- read.csv(file_path)

summarize_dataframe(df)


MyDF Summary Statistics ---
Shape:( 5076925 rows, 29 columns)
Mean: 5.753782 
Median: 0 
Standard Deviation: 20.61102 
---------------------------


# Project 2

## Question 3

In [ ]:
%%bash
cd /anvil/projects/tdm/data

pwd

echo $PWD

ls -p | grep -v /

ls -a /anvil/projects/tdm/data/movies_and_tv

cd

pwd

/anvil/projects/tdm/data
/anvil/projects/tdm/data
.
..
akas.csv
crew.csv
diagram.txt
episodes.csv
imdb-old.db
imdb.db
imdb2023
people.csv
ratings.csv
rotten_tomatoes
rotten_tomatoes_movies.csv
rotten_tomatoes_reviews.csv
the_office_dialogue.csv
titles.csv
/home/x-rdarshini

## Question 4

In [11]:
%%bash
ls -la /anvil/projects/tdm/opt/lmod

echo "Current directory: $(pwd)"

cd ../

echo "Current directory: $(pwd)"

total 15
drwxrwsr-x+ 15 7044182 7500002 4096 Aug 30 08:53 .
drwxrwsr-x+  6 7044182 7500002 4096 Aug 18 17:24 ..
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:05 aws
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:05 just
drwxrwsr-x+  2 7044182 7500002 4096 Jul  1  2022 libffi
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:06 linux_apps
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:05 lz4
drwxrwsr-x+  2 7044182 7500002 4096 Jul 15  2022 neofetch
drwxrwsr-x+  2 7044182 7500002 4096 Aug 25 14:01 python
drwxrwsr-x+  2 7044182 7500002 4096 Aug 19 10:59 r
drwxrwsr-x+  2 7044182 7500002 4096 Sep  8  2022 sqlite
drwxrwsr-x+  2 7044182 7500002 4096 Aug 30 08:53 test
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:06 vscode
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:05 xeyes
drwxrwsr-x+  2 7044182 7500002 4096 Aug 10 14:05 zstd
Current directory: /home/x-rdarshini/Our Project
Current directory: /home/x-rdarshini


Current directory: /home/x-rdarshini/Our Project
Current directory: /home/x-rdarshini

## Question 5

In [ ]:
%%bash

cd $SCRATCH_DIRECTORY
tokei $HOME
head -n 5 /anvil/datasets/training/anvil-101/batch-test/batch-test-README
tail -n 5 /anvil/datasets/training/anvil-101/batch-test/batch-test-README
wc -l < /anvil/datasets/training/anvil-101/batch-test/batch-test-README
stat -c %s /anvil/datasets/training/anvil-101/batch-test/batch-test-README
which tokei



## Question 6


In [ ]:
%%bash
cd $SCRATCH_DIRECTORY

cp /anvil/projects/tdm/data/movies_and_tv/imdb.db .

mkdir movies_and_tv

mv imdb.db movies_and_tv/

touch im_empty.txt

rm -r movies_and_tv

rm im_empty.txt




# Project 3

## Question 1

In [ ]:
%%bash
grep -Rin "朝阳区" /anvil/projects/tdm/data/air*

## Question 2

In [24]:
%%bash
head -1  /anvil/projects/tdm/data/consumer_complaints/processed.csv

Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Consumer consent provided?,Submitted via,Date sent to company,Company response to consumer,Timely response?,Consumer disputed?,Complaint ID


A. Date received,Product,Sub-product,Issue,Sub-issue,Consumer complaint narrative,Company public response,Company,State,ZIP code,Tags,Consumer consent provided?,Submitted via,Date sent to company,Company response to consumer,Timely response?,Consumer disputed?,Complaint ID

In [25]:
%%bash
cat /anvil/projects/tdm/data/consumer_complaints/processed.csv | head -n 6 | cut -d ',' -f 1,4,15

Date received,Issue,Company response to consumer
2013-08-26,Loan servicing payments escrow account,Closed with explanation
2023-07-23,Incorrect information on your report,In progress
2023-07-24,Attempts to collect debt not owed,Closed with explanation
2023-07-22,Problem with a credit reporting company's investigation into an existing problem,In progress
2023-07-21,Incorrect information on your report,Closed with explanation


B. Date received,Issue,Company response to consumer
2013-08-26,Loan servicing payments escrow account,Closed with explanation
2023-07-23,Incorrect information on your report,In progress
2023-07-24,Attempts to collect debt not owed,Closed with explanation
2023-07-22,Problem with a credit reporting company's investigation into an existing problem,In progress
2023-07-21,Incorrect information on your report,Closed with explanation


In [34]:
%%bash
cat /anvil/projects/tdm/data/consumer_complaints/processed.csv | grep -n 'fraudy fraudulent fraud' | cut -d ',' -f 1,4,6,15

3785027:2018-11-13,Managing the loan or lease,I financed my XXXX XXXX through Wells Fargo in XX/XX/XXXX on a five-year note. After I moved from Oklahoma to Wisconsin in XX/XX/XXXX I had a hiccup in my car insurance payment where it was sent to the old payment address that I'd used as an Oklahoma driver instead of the new payment address that I should have been using as a Wisconsin driver. That was totally my fault ; my insurance company contacted me the payment was applied to my insurance premium a few days after its original due date and I updated my online billpay to reflect the correct payment address for my future insurance premiums. nA few days later I got a letter from Wells Fargo stating that they'd been notified of a late payment by my insurance company and that if I didn't provide proof of coverage they were going to provide coverage themselves and bill me for the additional premiums. Now please don't get me wrong -- I was fine with that. It was my responsibility to insure the

3785027:2018-11-13,Managing the loan or lease,I financed my XXXX XXXX through Wells Fargo in XX/XX/XXXX on a five-year note. After I moved from Oklahoma to Wisconsin in XX/XX/XXXX I had a hiccup in my car insurance payment where it was sent to the old payment address that I'd used as an Oklahoma driver instead of the new payment address that I should have been using as a Wisconsin driver. That was totally my fault ; my insurance company contacted me the payment was applied to my insurance premium a few days after its original due date and I updated my online billpay to reflect the correct payment address for my future insurance premiums. nA few days later I got a letter from Wells Fargo stating that they'd been notified of a late payment by my insurance company and that if I didn't provide proof of coverage they were going to provide coverage themselves and bill me for the additional premiums. Now please don't get me wrong -- I was fine with that. It was my responsibility to insure the car that Wells Fargo was financing and it was my fault that an insurance premium was received late. I called my insurance agent provided him with the fax number for providing proof of insurance to Wells Fargo and they faxed the requested proof of insurance over while I was still on the line. nHere 's where things REALLY get interesting! nTwo or three days later I placed a follow-up call to Wells Fargo to ensure that they'd received what they needed. The woman that I spoke with sounded confused and said it looked like I'd been paying insurance premiums to Wells Fargo since I first bought my Jeep. I explained that this wasn't the case -- that I was calling in response to their letter -- and she explained that she was new didn't understand what she was seeing on my account but would get help from a supervisor.nI spent about eight to ten minutes on hold before she came back said she'd been mistaken that they'd received my fax and everything was fine. nNow one thing you should know is that I paid my car installments to Wells Fargo ahead of schedule. Every month I paid early and I always paid extra every month and so in XX/XX/XXXX I couldn't understand why the principal that I owed on my XXXX  was higher than it ought to have been if I'd paid my installments according to their schedule. Once again I called Wells Fargo. nThe woman that I spoke with this time said that my principal balance had gone up because Wells Fargo had billed me for insurance premiums. I asked to speak to a supervisor and suddenly her story changed : I was told that I'd been billed for insurance premiums but those premiums had also been refunded to me so that that in the end they hadn't increased my principal balance at all. I was told that I hadn't made as many extra payments as I thought I had and that their accounting was correct. Wells Fargo refused to provide me with any sort of billing and payment summary -- my questions weren't answered. I continued to pay off my XXXX early every month until the loan was paid off in XX/XX/XXXX. nIn XX/XX/XXXX I saw a news article about Wells Fargo being fined a XXXX dollars by the Consumer Financial Protection Bureau for fraudulently charging its car loan customers for unwanted auto insurance. nOf course I instantly knew that I'd been one of those customers even before I got the letter from Wells Fargo in XX/XX/XXXX. nThat letter was very apologetic and promised to make full restitution although Wells Fargo didn't use the word 'fraud ' the way that the CFPB had used it when fining the XXXX out of them. I think Wells Fargo 's letter used the word 'mistakenly ' or 'in error ' or 'inadvertently ' or somesuch nonsense sprinkled in with a few "" unfortunatelies '' and "" sincerelies. '' It was fraud. nPure and simple fraud. nFraudy fraudulent fraud. Wells Fargo fraudulently defrauded fraud victims with this sort of fraudy fraud. nI was asked to furnish Wells Fargo with 'proof ' that I'd maintained my own private auto insurance coverage so that Wells Fargo could reimburse me for any periods where I'd been charged for their in-house insurance product 'in error ' in order to qualify for a full refund ... along with Wells Fargo 's most profound apologies. Sincere apologies. Big wet crocodile teardrop apologies. Wells Fargo was SOOOOO very sorry that it had gotten caught stealing from me.nI sent Wells Fargo the requested information via USPS certified mail. I know they received it before the deadline. I have the proof of delivery.nAbout six months later I received a check from Wells Fargo.nIN.nTHE.nAMOUNT.nOF.nXXXX XXXX.nDOLLARS. nYou read that right. My 'full restitution ' for Wells Fargo 's mistaken completely in error unfortunate deeply regretted certainly-not-a-case-of-fraud was {$29.00}. nKeep in mind that I still haven't received any full accounting of my account history because Wells Fargo won't provide that information to me. nAfter working the phones for a few days I finally got in touch with Wells Fargo 's office of the president. What I learned was that Wells Fargo had fraudulently added {$840.00} to my principal loan balance in XX/XX/XXXX and XX/XX/XXXX long before I moved to Wisconsin and had my little incident with a misapplied insurance payment. I was charged another lump-sum amount for insurance coverage in XX/XX/XXXX and XX/XX/XXXX but that they'd canceled that coverage after I called them to make sure they're received my agent 's fax refunding all but {$260.00} of that second amount.nKeep in mind the agent who presumably canceled that insurance in XX/XX/XXXX told me that she'd been mistaken about seeing insurance on my account. nKeep in mind that I'd been told by a supervisor that I'd been completely refunded for all insurance premiums so that they hadn't increased my loan amount. nKeep in mind that Wells Fargo initially tried to make me go away with a {$29.00} check. nThat's the sort of fraudy fraudulent fraud that Wells Fargo defrauds its fraud-victim customers with. Fraudulently. nFraud. Fraud. Fraud.nBy my reckoning then Wells Fargo stole {$840.00} and {$260.00} from me plus compounded interest on this amount from XX/XX/XXXX to XX/XX/XXXX less the {$29.00} insult check that they sent to me. nKeep in mind that these are Wells Fargo 's numbers and that to this day I don't know exactly what charges were applied to my principal loan amount or on what dates or how much interest was charged on this balance because Wells Fargo won't tell me. nCompletely setting aside interest amounts -- and I want a refund on interest paid I just don't know how much that was -- Wells Fargo owes me {$1000.00} in fraudulent charges applied to my principal loan amount. I think we've been more or less in agreement on that amount since XX/XX/XXXX -- or at least Wells Fargo doesn't dispute that amount -- but they still won't pay me. nThe funny thing is that Wells Fargo has spent about XXXX months "" verifying my insurance coverage '' prior to payment -- that strikes me as funny because they really should have done that BEFORE billing me for their insurance product. nLate in XX/XX/XXXX I read in the news that Wells Fargo had told the CFPB that it wouldn't be prepared to begin making settlement offers to its victims until XX/XX/XXXX and that it wouldn't complete its restitution payments until XX/XX/XXXX. nSo as I understand it Wells Fargo notified me of fraudulent charges on my account in XX/XX/XXXX five years after it added those charges to to my loan balance and four years after it first lied to me about those charges -- and now Wells Fargo has decided that it's going to take another two years to decide how to make that situation right as if that solution weren't already obvious and THEN Wells Fargo is going to graciously allow itself another year to actually return the money that it had no right to take in the first place. nThe REAL kicker was reading last week that a top official at Wells Fargo Bank was aware of this fraud in XX/XX/XXXX but did nothing to stop it -- and HE STILL HAS A JOB AT THIS COMPANY. nI just want my money back. How in God 's name do I push Wells Fargo 's monstrous bureaucracy into action?,Closed with monetary relief

## Question 3

In [19]:
%%bash
cat /anvil/projects/tdm/data/consumer_complaints/processed.csv | grep ",IN,\|,OH,\|,IL,\|,WI,\|,MI," |cut -d ',' -f 1,4,6,15,9 > midwest.csv

In [20]:
%%bash
row_count=$(wc -l < midwest.csv)

file_size_bytes=$(ls -l midwest.csv | cut -d ' ' -f 5)
file_size_mb=$((file_size_bytes / 1024 / 1024))

echo "Number of rows: $row_count"
echo "File size (MB): $file_size_mb MB"


Number of rows: 416632
File size (MB): 192 MB


## Question 4

In [35]:
%%bash
grep -iE 'improper|struggling|incorrect' midwest.csv | wc -l


181277


181277

## Question 5

In [22]:
%%bash
awk -F ',' '{print $1}' /anvil/projects/tdm/data/consumer_complaints/processed.csv | sort | uniq -c | sort -nr | head -1


   4773 2023-06-27


   4773 2023-06-27

In [23]:
%%bash
cut -d ',' -f 3 /anvil/projects/tdm/data/consumer_complaints/processed.csv | sort | uniq -c


 235291 
   3417 (CD) Certificate of deposit
   3754 Auto
  14096 Auto debt
   3828 CD (Certificate of Deposit)
    625 Cashing a check without an account
    266 Check cashing
    919 Check cashing service
 198099 Checking account
  25380 Conventional adjustable mortgage (ARM)
  70607 Conventional fixed mortgage
  93711 Conventional home mortgage
  28692 Credit card
  82020 Credit card debt
    102 Credit repair
   4794 Credit repair services
2039707 Credit reporting
   1247 Debt settlement
  17879 Domestic (US) money transfer
     12 Electronic Benefit Transfer / EBT card
  48176 FHA mortgage
   2474 Federal student loan
   3785 Federal student loan debt
  38842 Federal student loan servicing
    647 Foreign currency exchange
   1752 General purpose card
 152378 General-purpose credit card or charge card
   8235 General-purpose prepaid card
    964 Gift card
    402 Gift or merchant card
   8880 Government benefit card
    389 Government benefit payment card
  11623 Home equity loan 

235291 
   3417 (CD) Certificate of deposit
   3754 Auto
  14096 Auto debt
   3828 CD (Certificate of Deposit)
    625 Cashing a check without an account
    266 Check cashing
    919 Check cashing service
 198099 Checking account
  25380 Conventional adjustable mortgage (ARM)
  70607 Conventional fixed mortgage
  93711 Conventional home mortgage
  28692 Credit card
  82020 Credit card debt
    102 Credit repair
   4794 Credit repair services
2039707 Credit reporting
   1247 Debt settlement
  17879 Domestic (US) money transfer
     12 Electronic Benefit Transfer / EBT card
  48176 FHA mortgage
   2474 Federal student loan
   3785 Federal student loan debt
  38842 Federal student loan servicing
    647 Foreign currency exchange
   1752 General purpose card
 152378 General-purpose credit card or charge card
   8235 General-purpose prepaid card
    964 Gift card
    402 Gift or merchant card
   8880 Government benefit card
    389 Government benefit payment card
  11623 Home equity loan or line of credit
   8961 Home equity loan or line of credit (HELOC)
 110166 I do not know
    207 ID prepaid card
  21669 Installment loan
  10067 International money transfer
   7117 Lease
  38448 Loan
  21184 Medical
  53138 Medical debt
  20603 Mobile or digital wallet
    418 Mobile wallet
    900 Money order
   4809 Mortgage
   6300 Mortgage debt
  25162 Non-federal student loan
  44524 Other (i.e. phone health club etc.)
  17825 Other bank product/service
  18759 Other banking product or service
  97496 Other debt
  86633 Other mortgage
  15590 Other personal consumer report
    191 Other special purpose card
  15484 Other type of mortgage
    161 Pawn loan
  14339 Payday loan
   9203 Payday loan debt
   1289 Payroll card
   8953 Personal line of credit
  14192 Private student loan
   3821 Private student loan debt
    672 Refund anticipation check
   4389 Reverse mortgage
  18212 Savings account
    662 Second mortgage
  27232 Store credit card
     31 Student prepaid card
      1 Sub-product
   3272 Title loan
     37 Transit card
    948 Traveler's check or cashier's check
     88 Traveler’s/Cashier’s checks
  15217 VA mortgage
   2655 Vehicle lease
  17779 Vehicle loan
   8769 Virtual currency

# Project 4

## Question 1

In [1]:
%%bash
cat /anvil/projects/tdm/data/stackoverflow/unprocessed/2011.csv | head -n10 | tr ',' '\n' | wc -w

132395


132395

## Question 2

In [1]:
%%bash
head /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" 'NR==5{print NF}'

24


24

## Question 3

In [ ]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F ';' '{print $20"  " $16}' | sort -nr | head -n 6

27654  GLENGOYNE SINGLE MALT SCOTCH 50YR DECANTER
11520  CEDAR RIDGE RESERVE BOURBON PRIVATE CASK DNO
9702  CEDAR RIDGE RYE WHISKEY - WHOLE CASK BUY
9702  CEDAR RIDGE RYE WHISKEY - WHOLE CASK BUY
9150  CEDAR RIDGE MALTED RYE BARREL URBANDALE
9150  CEDAR RIDGE MALTED RYE BARREL MARION


## Question 4

In [ ]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | cut -d';' -f18 | sort -n | uniq -c

10 0
      1 Bottle Volume (ml)
      1 12
      1 15
   2209 20
    253 25
1145429 50
 213852 100
    793 150
     54 175
 944624 200
    876 250
  94482 300
    130 301
      7 355
2784582 375
   2223 400
     26 425
 730575 500
    349 502
 221259 600
    206 603
   4331 700
12135806 750
  23572 800
   3947 850
   1521 900
    973 950
3308611 1000
    511 1125
  20363 1200
     63 1420
    190 1500
      3 1508
5289184 1750
   5981 1800
   1896 1850
    566 1950
     23 2000
     16 2125
   2319 2250
   8843 2400
     42 2550
  95623 3000
   1279 3500
    161 3600
     18 3900
    326 4500
    476 4800
    194 5250
   1299 6000
      1 7580
      3 9000
     24 31500
      8 140000
      2 180000
      2 189000
      1 225000
     24 378000

Explanation: cut -d';' -f18 takes the 18th column of data which is the bottle's volume, sort-n sorts it in ascending order and uniq -c tells how many times a line was repeated(in this case the integer of volume) by displaying a number as a prefix with the line.

# Project 5

## Question 1

A.

In [4]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | grep "WHISKIES" | awk -F";" '{myprices += $19} END{print myprices}'

8.59911e+07


8.59911e+07

B.

In [5]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | grep "CEDAR RAPIDS" | awk -F";" '{myprices += $19} END{print myprices}'

1.7924e+07


1.7924e+07

## Question 2

A.

In [6]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{print $4}' | sort | uniq -c | sort -n | tail -1

 223447 HY-VEE #3 / BDI / DES MOINES


223447 HY-VEE #3 / BDI / DES MOINES

B.

In [7]:
%%bash
grep 'HY-VEE #3 / BDI / DES MOINES' /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{total += $19} END {print total}'

2.88549e+06


2.88549e+06

## Question 3

A.

In [8]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{total += $23} END {print total}'

2.47828e+08


2.47828e+08

B.

In [9]:
%%bash
grep 'VODKA 80 PROOF' /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{total += $23} END {print total}'

1.78615e+07


1.78615e+07

## Question 4

A.

In [10]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{print $5, $6, $7}' | sort | uniq -c | sort -n | tail -1


 223447 3221 SE 14TH ST DES MOINES 50320


 223447 3221 SE 14TH ST DES MOINES 50320

B. Yes. They agree. Since both store's frequncy is the highest, one prints the storename, and other prints the address of it.

## Question 5

In [11]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" 'NR > 1 {category[$12] += $21} END {for (cat in category) print cat, category[cat]}'

TEQUILA 4079513
VODKA 80 PROOF 18124615
BLENDED WHISKIES 10411543
TRIPLE SEC 2167823
TROPICAL FRUIT SCHNAPPS 60062
IMPORTED GRAPE BRANDIES 1576061
DARK CREME DE CACAO 32714
IMPORTED WHISKIES 72
 252120
PEACH SCHNAPPS 567257
IMPORTED AMARETTO 86235
IMPORTED DRY GINS 2173057
TEMPORARY & SPECIALTY PACKAGES 2366157
DELISTED ITEMS 149
AMERICAN COCKTAILS 2120834
DELISTED / SPECIAL ORDER ITEMS 12
CINNAMON SCHNAPPS 179177
APRICOT BRANDIES 256932
SINGLE BARREL BOURBON WHISKIES 204867
AMERICAN FLAVORED VODKA 6648136
AMERICAN SCHNAPPS 4842369
OTHER PROOF VODKA 16618
HIGH PROOF BEER - AMERICAN 54
CREME DE ALMOND 6620
AMERICAN GRAPE BRANDIES 2630010
SINGLE MALT SCOTCH 896395
CANADIAN WHISKIES 31529658
DECANTERS & SPECIALTY PACKAGES 620739
IMPORTED FLAVORED VODKA 2406713
MISC. IMPORTED CORDIALS & LIQUEURS 1877575
BOTTLED IN BOND BOURBON 154984
WATERMELON SCHNAPPS 156548
AMERICAN SLOE GINS 57298
IMPORTED DISTILLED SPIRITS SPECIALTY 1536183
WHISKEY LIQUEUR 24023352
STRAIGHT RYE WHISKIES 1409518
IMPORT

# Project 6

## Question 1

In [10]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F"|" '{print NF}' | uniq -c | awk -F" " '{print "rows: "$1; print "columns: "$2}'

rows: 135304
columns: 1


rows: 135304
columns: 1

## Question 2

In [11]:
%%bash
head /anvil/projects/tdm/data/restaurant/orders.csv | awk -F"," '{print $25}'| sort | uniq -c | grep -v LOCATION_TYPE | awk 'BEGIN{print "Location Type   Number of Orders"; print "-------------   ----------------"; } { printf " %-10s%6s\n", $2, $1}'  

Location Type   Number of Orders
-------------   ----------------
                2
 Home           1
 Other          1
 Work           5


Location Type   Number of Orders
-------------   ----------------
                2
 Home           1
 Other          1
 Work           5

## Question 3

In [12]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F"," '{print $21}'| awk -F"-" '{print $1}' | sort | uniq | grep -v delivery_date


2019


2019

## Question 4

In [3]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F"," '{total+= $4} END {printf "Year Summary of Orders in dollars\n2019 $%.2f", total}'

Year Summary of Orders in dollars
2019 $2080217.70

Markdown notes and sentences and analysis written here.

## Question 5

In [36]:
%%bash
cat /anvil/projects/tdm/data/whin/observations.csv | awk -F"|" '{print $3}' | awk -F"-" '{print $1,$2}' | grep -v observation_time | sort | uniq -c | awk 'BEGIN{print "Month and Year   Number of Occurances"; print "--------------   ---------------------"; } { printf "%s-%s%16s\n", $2, $3, $1}'  

Month and Year   Number of Occurances
--------------   ---------------------
2019-07            2002
2019-08            3969
2019-09            5749
2019-10            5948
2019-11            5731
2019-12            5918
2020-01            5482
2020-02            5008
2020-03            5436
2020-04           19326
2020-05           42819
2020-06           45557
2020-07           50557
2020-08           53288
2020-09           53400
2020-10           58283
2020-11           57501
2020-12           59509
2021-01           61254
2021-02           60306
2021-03           81536
2021-04           80116
2021-05           83307
2021-06           80454
2021-07           67544


Month and Year   Number of Occurances
--------------   ---------------------
2019-07            2002
2019-08            3969
2019-09            5749
2019-10            5948
2019-11            5731
2019-12            5918
2020-01            5482
2020-02            5008
2020-03            5436
2020-04           19326
2020-05           42819
2020-06           45557
2020-07           50557
2020-08           53288
2020-09           53400
2020-10           58283
2020-11           57501
2020-12           59509
2021-01           61254
2021-02           60306
2021-03           81536
2021-04           80116
2021-05           83307
2021-06           80454
2021-07           67544

# Project 7

## Question 1

In [2]:
%%bash
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{mytotal[$4] += $22} END{for(i in mytotal) {print mytotal[i], i}}' | sort -g |tail

2.68169e+07 CENTRAL CITY LIQUOR, INC.
2.79315e+07 SAM'S CLUB 8238 / DAVENPORT
3.09229e+07 LOT-A-SPIRITS
3.38266e+07 WILKIE LIQUORS
3.42666e+07 BENZ DISTRIBUTING
3.96246e+07 SAM'S CLUB 6344 / WINDSOR HEIGHTS
4.13889e+07 SAM'S CLUB 8162 / CEDAR RAPIDS
4.87686e+07 HY-VEE WINE AND SPIRITS / IOWA CITY
1.08328e+08 CENTRAL CITY 2
1.23313e+08 HY-VEE #3 / BDI / DES MOINES


2.68169e+07 CENTRAL CITY LIQUOR, INC.
2.79315e+07 SAM'S CLUB 8238 / DAVENPORT
3.09229e+07 LOT-A-SPIRITS
3.38266e+07 WILKIE LIQUORS
3.42666e+07 BENZ DISTRIBUTING
3.96246e+07 SAM'S CLUB 6344 / WINDSOR HEIGHTS
4.13889e+07 SAM'S CLUB 8162 / CEDAR RAPIDS
4.87686e+07 HY-VEE WINE AND SPIRITS / IOWA CITY
1.08328e+08 CENTRAL CITY 2
1.23313e+08 HY-VEE #3 / BDI / DES MOINES

## Question 2

In [6]:
%%bash
cat /anvil/projects/tdm/data/election/escaped2020sample.txt | sed 's/"//g' | awk -F"|" '{mytotal[$9","$10] += $15} END{for(i in mytotal) {print mytotal[i], i}}' | sort -n | tail

7115650 SEATTLE,WA
7141597 ATLANTA,GA
8287423 BOSTON,MA
11032607 CHICAGO,IL
11642460 DALLAS,TX
12463126 HOUSTON,TX
14141160 LOS ANGELES,CA
37152718 WASHINGTON,DC
38701698 SAN FRANCISCO,CA
331576505 NEW YORK,NY


7115650 SEATTLE,WA
7141597 ATLANTA,GA
8287423 BOSTON,MA
11032607 CHICAGO,IL
11642460 DALLAS,TX
12463126 HOUSTON,TX
14141160 LOS ANGELES,CA
37152718 WASHINGTON,DC
38701698 SAN FRANCISCO,CA
331576505 NEW YORK,NY

## Question 3

In [13]:
%%bash
cut -d ',' -f 6 /anvil/projects/tdm/data/flights/1990.csv | grep -v FlightDate | awk -F '-' '{print $2 "/" $1}' | sort | uniq -c

 442700 01/1990
 396926 02/1990
 444320 03/1990
 433420 04/1990
 447371 05/1990
 437395 06/1990
 455057 07/1990
 457886 08/1990
 435638 09/1990
 455047 10/1990
 428781 11/1990
 436352 12/1990


442700 01/1990
 396926 02/1990
 444320 03/1990
 433420 04/1990
 447371 05/1990
 437395 06/1990
 455057 07/1990
 457886 08/1990
 435638 09/1990
 455047 10/1990
 428781 11/1990
 436352 12/1990

## Question 4

In [26]:
%%bash
{ echo "flight_month_year;total_number_of_flights";cut -d ',' -f 6 /anvil/projects/tdm/data/flights/1990.csv | awk -F '-' '{print $2 "/" $1}' | sort -n | uniq -c | grep -v FlightDate | awk '{print $2 ";" $1}' | sort -t ';' -k 2n; } > 1990_flight_info.csv

# Project 8

## Question 1

In [40]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [41]:
%sql SELECT tbl_name FROM sqlite_master WHERE type='table';

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('teamsfranchises',),
 ('teams',),
 ('teamshalf',),
 ('people',),
 ('allstarfull',),
 ('appearances',),
 ('awardsmanagers',),
 ('awardsplayers',),
 ('awardssharemanagers',),
 ('awardsshareplayers',),
 ('batting',),
 ('battingpost',),
 ('schools',),
 ('collegeplaying',),
 ('fielding',),
 ('fieldingof',),
 ('fieldingofsplit',),
 ('fieldingpost',),
 ('halloffame',),
 ('parks',),
 ('homegames',),
 ('managers',),
 ('managershalf',),
 ('pitching',),
 ('pitchingpost',),
 ('salaries',),
 ('seriespost',)]

tbl_name
teamsfranchises
teams
teamshalf
people
allstarfull
appearances
awardsmanagers
awardsplayers
awardssharemanagers
awardsshareplayers
batting
battingpost
schools
collegeplaying
fielding
fieldingof
fieldingofsplit
fieldingpost
halloffame
parks
homegames
managers
managershalf
pitching
pitchingpost
salaries
seriespost


## Question 2

A.

In [42]:
%%sql
SELECT * FROM people LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('aardsda01', 1981, 12, 27, 'USA', 'CO', 'Denver', '', '', '', '', '', '', 'David', 'Aardsma', 'David Allan', 215, 75, 'R', 'R', '2004-04-06', '2015-08-23', 'aardd001', 'aardsda01'),
 ('aaronha01', 1934, 2, 5, 'USA', 'AL', 'Mobile', 2021, 1, 22, 'USA', 'GA', 'Atlanta', 'Hank', 'Aaron', 'Henry Louis', 180, 72, 'R', 'R', '1954-04-13', '1976-10-03', 'aaroh101', 'aaronha01'),
 ('aaronto01', 1939, 8, 5, 'USA', 'AL', 'Mobile', 1984, 8, 16, 'USA', 'GA', 'Atlanta', 'Tommie', 'Aaron', 'Tommie Lee', 190, 75, 'R', 'R', '1962-04-10', '1971-09-26', 'aarot101', 'aaronto01'),
 ('aasedo01', 1954, 9, 8, 'USA', 'CA', 'Orange', '', '', '', '', '', '', 'Don', 'Aase', 'Donald William', 190, 75, 'R', 'R', '1977-07-26', '1990-10-03', 'aased001', 'aasedo01'),
 ('abadan01', 1972, 8, 25, 'USA', 'FL', 'Palm Beach', '', '', '', '', '', '', 'Andy', 'Abad', 'Fausto Andres', 184, 73, 'L', 'L', '2001-09-10', '2006-04-13', 'abada001', 'abadan01')]

playerID	birthYear	birthMonth	birthDay	birthCountry	birthState	birthCity	deathYear	deathMonth	deathDay	deathCountry	deathState	deathCity	nameFirst	nameLast	nameGiven	weight	height	bats	throws	debut	finalGame	retroID	bbrefID
aardsda01	1981	12	27	USA	CO	Denver							David	Aardsma	David Allan	215	75	R	R	2004-04-06	2015-08-23	aardd001	aardsda01
aaronha01	1934	2	5	USA	AL	Mobile	2021	1	22	USA	GA	Atlanta	Hank	Aaron	Henry Louis	180	72	R	R	1954-04-13	1976-10-03	aaroh101	aaronha01
aaronto01	1939	8	5	USA	AL	Mobile	1984	8	16	USA	GA	Atlanta	Tommie	Aaron	Tommie Lee	190	75	R	R	1962-04-10	1971-09-26	aarot101	aaronto01
aasedo01	1954	9	8	USA	CA	Orange							Don	Aase	Donald William	190	75	R	R	1977-07-26	1990-10-03	aased001	aasedo01
abadan01	1972	8	25	USA	FL	Palm Beach							Andy	Abad	Fausto Andres	184	73	L	L	2001-09-10	2006-04-13	abada001	abadan01

In [43]:
%%sql
SELECT * FROM batting LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abercda01', 1871, 1, 'TRO', 'NA', 1, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '', 0),
 ('addybo01', 1871, 1, 'RC1', 'NA', 25, 118, 30, 32, 6, 0, 0, 13, 8, 1, 4, 0, '', '', '', '', 0),
 ('allisar01', 1871, 1, 'CL1', 'NA', 29, 137, 28, 40, 4, 5, 0, 19, 3, 1, 2, 5, '', '', '', '', 1),
 ('allisdo01', 1871, 1, 'WS3', 'NA', 27, 133, 28, 44, 10, 2, 2, 27, 1, 1, 0, 2, '', '', '', '', 0),
 ('ansonca01', 1871, 1, 'RC1', 'NA', 25, 120, 29, 39, 11, 3, 0, 16, 6, 2, 2, 1, '', '', '', '', 0)]

playerID	yearID	stint	teamID	lgID	G	AB	R	H	2B	3B	HR	RBI	SB	CS	BB	SO	IBB	HBP	SH	SF	GIDP
abercda01	1871	1	TRO	NA	1	4	0	0	0	0	0	0	0	0	0	0					0
addybo01	1871	1	RC1	NA	25	118	30	32	6	0	0	13	8	1	4	0					0
allisar01	1871	1	CL1	NA	29	137	28	40	4	5	0	19	3	1	2	5					1
allisdo01	1871	1	WS3	NA	27	133	28	44	10	2	2	27	1	1	0	2					0
ansonca01	1871	1	RC1	NA	25	120	29	39	11	3	0	16	6	2	2	1					0


In [44]:
%%sql
SELECT * FROM fielding LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abercda01', 1871, 1, 'TRO', 'NA', 'SS', 1, 1, 24, 1, 3, 2, 0, '', '', '', '', ''),
 ('addybo01', 1871, 1, 'RC1', 'NA', '2B', 22, 22, 606, 67, 72, 42, 5, '', '', '', '', ''),
 ('addybo01', 1871, 1, 'RC1', 'NA', 'SS', 3, 3, 96, 8, 14, 7, 0, '', '', '', '', ''),
 ('allisar01', 1871, 1, 'CL1', 'NA', '2B', 2, 0, 18, 1, 4, 0, 0, '', '', '', '', ''),
 ('allisar01', 1871, 1, 'CL1', 'NA', 'OF', 29, 29, 729, 51, 3, 7, 1, '', '', '', '', '')]

playerID	yearID	stint	teamID	lgID	POS	G	GS	InnOuts	PO	A	E	DP	PB	WP	SB	CS	ZR
abercda01	1871	1	TRO	NA	SS	1	1	24	1	3	2	0					
addybo01	1871	1	RC1	NA	2B	22	22	606	67	72	42	5					
addybo01	1871	1	RC1	NA	SS	3	3	96	8	14	7	0					
allisar01	1871	1	CL1	NA	2B	2	0	18	1	4	0	0					
allisar01	1871	1	CL1	NA	OF	29	29	729	51	3	7	1					


In [45]:
%%sql
SELECT * FROM managers LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('wrighha01', 1871, 'BS1', 'NA', 1, 31, 20, 10, 3, 'Y'),
 ('woodji01', 1871, 'CH1', 'NA', 1, 28, 19, 9, 2, 'Y'),
 ('paborch01', 1871, 'CL1', 'NA', 1, 29, 10, 19, 8, 'Y'),
 ('lennobi01', 1871, 'FW1', 'NA', 1, 14, 5, 9, 8, 'Y'),
 ('deaneha01', 1871, 'FW1', 'NA', 2, 5, 2, 3, 8, 'Y')]

playerID	yearID	teamID	lgID	inseason	G	W	L	teamRank	plyrMgr
wrighha01	1871	BS1	NA	1	31	20	10	3	Y
woodji01	1871	CH1	NA	1	28	19	9	2	Y
paborch01	1871	CL1	NA	1	29	10	19	8	Y
lennobi01	1871	FW1	NA	1	14	5	9	8	Y
deaneha01	1871	FW1	NA	2	5	2	3	8	Y


In [46]:
%%sql
SELECT * FROM teamsfranchises LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('ALT', 'Altoona Mountain City', 'N', ''),
 ('ANA', 'Los Angeles Angels of Anaheim', 'Y', ''),
 ('ARI', 'Arizona Diamondbacks', 'Y', ''),
 ('ATH', 'Philadelphia Athletics', 'N', 'PNA'),
 ('ATL', 'Atlanta Braves', 'Y', 'BNA')]

franchID	franchName	active	NAassoc
ALT	Altoona Mountain City	N	
ANA	Los Angeles Angels of Anaheim	Y	
ARI	Arizona Diamondbacks	Y	
ATH	Philadelphia Athletics	N	PNA
ATL	Atlanta Braves	Y	BNA


In [47]:
%%sql
SELECT * FROM collegeplaying LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('aardsda01', 'pennst', 2001),
 ('aardsda01', 'rice', 2002),
 ('aardsda01', 'rice', 2003),
 ('abadan01', 'gamiddl', 1992),
 ('abadan01', 'gamiddl', 1993)]

playerID	schoolID	yearID
aardsda01	pennst	2001
aardsda01	rice	2002
aardsda01	rice	2003
abadan01	gamiddl	1992
abadan01	gamiddl	1993

B.

In [48]:
%%sql
SELECT COUNT(*) FROM people LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(20676,)]

20676

In [49]:
%%sql
SELECT COUNT(*) FROM batting LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(112184,)]

112184

In [50]:
%%sql
SELECT COUNT(*) FROM fielding LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(149365,)]

149365

In [51]:
%%sql
SELECT COUNT(*) FROM managers LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(3718,)]

3718

In [52]:
%%sql
SELECT COUNT(*) FROM teamsfranchises LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(120,)]

120

In [53]:
%%sql
SELECT COUNT(*) FROM collegeplaying LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(17350,)]

17350

## Question 3

In [54]:
%%sql
SELECT * FROM people WHERE nameFirst = 'Mike' AND nameLast = 'Trout' LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('troutmi01', 1991, 8, 7, 'USA', 'NJ', 'Vineland', '', '', '', '', '', '', 'Mike', 'Trout', 'Michael Nelson', 235, 74, 'R', 'R', '2011-07-08', '2022-10-05', 'troum001', 'troutmi01')]

playerID	birthYear	birthMonth	birthDay	birthCountry	birthState	birthCity	deathYear	deathMonth	deathDay	deathCountry	deathState	deathCity	nameFirst	nameLast	nameGiven	weight	height	bats	throws	debut	finalGame	retroID	bbrefID
troutmi01	1991	8	7	USA	NJ	Vineland							Mike	Trout	Michael Nelson	235	74	R	R	2011-07-08	2022-10-05	troum001	troutmi01


## Question 4

A.

In [57]:
%%sql
SELECT playerID, yearID, HR FROM batting where playerID ='troutmi01'

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('troutmi01', 2011, 5),
 ('troutmi01', 2012, 30),
 ('troutmi01', 2013, 27),
 ('troutmi01', 2014, 36),
 ('troutmi01', 2015, 41),
 ('troutmi01', 2016, 29),
 ('troutmi01', 2017, 33),
 ('troutmi01', 2018, 39),
 ('troutmi01', 2019, 45),
 ('troutmi01', 2020, 17),
 ('troutmi01', 2021, 8),
 ('troutmi01', 2022, 40)]

playerID	yearID	HR
troutmi01	2011	5
troutmi01	2012	30
troutmi01	2013	27
troutmi01	2014	36
troutmi01	2015	41
troutmi01	2016	29
troutmi01	2017	33
troutmi01	2018	39
troutmi01	2019	45
troutmi01	2020	17
troutmi01	2021	8
troutmi01	2022	40

B.

In [59]:
%%sql
SELECT playerID, teamID, HR FROM batting WHERE playerID ='troutmi01' AND yearID = '2022';

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('troutmi01', 'LAA', 40)]

playerID	teamID	HR
troutmi01	LAA	40

## Question 5

A.

In [60]:
%%sql
SELECT * FROM people WHERE nameFirst = 'Albert' AND nameLast = 'Abreu' LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abreual01', 1995, 9, 26, 'D.R.', 'Monte Cristi', 'Guayubin', '', '', '', '', '', '', 'Albert', 'Abreu', 'Albert Enmanuel', 190, 74, 'R', 'R', '2020-08-08', '2022-10-05', 'abrea001', 'abreual01')]

playerID	birthYear	birthMonth	birthDay	birthCountry	birthState	birthCity	deathYear	deathMonth	deathDay	deathCountry	deathState	deathCity	nameFirst	nameLast	nameGiven	weight	height	bats	throws	debut	finalGame	retroID	bbrefID
abreual01	1995	9	26	D.R.	Monte Cristi	Guayubin							Albert	Abreu	Albert Enmanuel	190	74	R	R	2020-08-08	2022-10-05	abrea001	abreual01

B.

In [61]:
%%sql
SELECT playerID, teamID, HR FROM batting where playerID ='abreual01' AND yearID = '2022'

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abreual01', 'TEX', 0), ('abreual01', 'KCA', 0), ('abreual01', 'NYA', 0)]

playerID	teamID	HR
abreual01	TEX	0
abreual01	KCA	0
abreual01	NYA	0


C.

In [62]:
%%sql
SELECT b.playerID, nameLast, nameFirst, weight, height, birthYear, HR, teamID, yearID FROM batting AS b JOIN people AS p ON b.playerID = p.playerID WHERE p.playerID ='abreual01' AND yearID = '2022'

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abreual01', 'Abreu', 'Albert', 190, 74, 1995, 0, 'TEX', 2022),
 ('abreual01', 'Abreu', 'Albert', 190, 74, 1995, 0, 'KCA', 2022),
 ('abreual01', 'Abreu', 'Albert', 190, 74, 1995, 0, 'NYA', 2022)]

playerID	nameLast	nameFirst	weight	height	birthYear	HR	teamID	yearID
abreual01	Abreu	Albert	190	74	1995	0	TEX	2022
abreual01	Abreu	Albert	190	74	1995	0	KCA	2022
abreual01	Abreu	Albert	190	74	1995	0	NYA	2022


# Project 9

## Question 1

A.

In [2]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [3]:
%%sql
SELECT * FROM seriespost WHERE yearID = 2022 AND round = 'WS';

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(2022, 'WS', 'HOU', 'NL', 'PHI', 'NL', 4, 2, 0)]

yearID	round	teamIDwinner	lgIDwinner	teamIDloser	lgIDloser	wins	losses	ties
2022	WS	HOU	NL	PHI	NL	4	2	0


B.

In [4]:
%%sql
SELECT ROUND(CAST(SUM(HR) AS REAL)/SUM(H), 2) FROM batting WHERE teamID = 'HOU' AND yearID = 2022

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(0.16,)]

0.16

In [5]:
%%sql
SELECT ROUND(CAST(SUM(RBI) AS REAL)/SUM(H), 2) FROM batting WHERE teamID = 'HOU' AND yearID = 2022

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(0.53,)]

0.53

## Question 2

A.

In [6]:
%%sql
SELECT b.playerID, b.RBI FROM batting AS b JOIN people AS p ON b.playerID = p.playerID WHERE teamID = 'HOU' AND yearID = 2022 ORDER BY b.RBI ASC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abreubr01', 0),
 ('baezpe01', 0),
 ('bielabr01', 0),
 ('blancro01', 0),
 ('brownhu01', 0),
 ('garcilu05', 0),
 ('javiecr01', 0),
 ('jonesta01', 0),
 ('martise01', 0),
 ('matonph01', 0),
 ('mcculla02', 0),
 ('montera01', 0),
 ('mushipa01', 0),
 ('nerishe01', 0),
 ('odorija01', 0),
 ('pareden01', 0),
 ('perezjo01', 0),
 ('pressry01', 0),
 ('smithwi04', 0),
 ('stanery01', 0),
 ('taylobl01', 0),
 ('urquijo01', 0),
 ('valdefr01', 0),
 ('verlaju01', 0),
 ('diazya02', 1),
 ('goodrni01', 1),
 ('castrja01', 3),
 ('leeko01', 4),
 ('henslda01', 5),
 ('matijjj01', 5),
 ('sirijo01', 10),
 ('vazquch01', 10),
 ('meyerja02', 15),
 ('dubonma01', 16),
 ('mancitr01', 22),
 ('brantmi02', 26),
 ('diazal02', 38),
 ('mccorch01', 44),
 ('maldoma01', 45),
 ('gourryu01', 53),
 ('altuvjo01', 57),
 ('penaje02', 63),
 ('bregmal01', 93),
 ('alvaryo01', 97),
 ('tuckeky01', 107)]

playerID	RBI
abreubr01	0
baezpe01	0
bielabr01	0
blancro01	0
brownhu01	0
garcilu05	0
javiecr01	0
jonesta01	0
martise01	0
matonph01	0
mcculla02	0
montera01	0
mushipa01	0
nerishe01	0
odorija01	0
pareden01	0
perezjo01	0
pressry01	0
smithwi04	0
stanery01	0
taylobl01	0
urquijo01	0
valdefr01	0
verlaju01	0
diazya02	1
goodrni01	1
castrja01	3
leeko01	4
henslda01	5
matijjj01	5
sirijo01	10
vazquch01	10
meyerja02	15
dubonma01	16
mancitr01	22
brantmi02	26
diazal02	38
mccorch01	44
maldoma01	45
gourryu01	53
altuvjo01	57
penaje02	63
bregmal01	93
alvaryo01	97
tuckeky01	107

B.

In [7]:
%%sql
SELECT b.playerID, b.RBI FROM batting AS b JOIN people AS p ON b.playerID = p.playerID WHERE teamID = 'HOU' AND yearID = 2022 ORDER BY b.RBI DESC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('tuckeky01', 107),
 ('alvaryo01', 97),
 ('bregmal01', 93),
 ('penaje02', 63),
 ('altuvjo01', 57),
 ('gourryu01', 53),
 ('maldoma01', 45),
 ('mccorch01', 44),
 ('diazal02', 38),
 ('brantmi02', 26),
 ('mancitr01', 22),
 ('dubonma01', 16),
 ('meyerja02', 15),
 ('sirijo01', 10),
 ('vazquch01', 10),
 ('henslda01', 5),
 ('matijjj01', 5),
 ('leeko01', 4),
 ('castrja01', 3),
 ('diazya02', 1),
 ('goodrni01', 1),
 ('abreubr01', 0),
 ('baezpe01', 0),
 ('bielabr01', 0),
 ('blancro01', 0),
 ('brownhu01', 0),
 ('garcilu05', 0),
 ('javiecr01', 0),
 ('jonesta01', 0),
 ('martise01', 0),
 ('matonph01', 0),
 ('mcculla02', 0),
 ('montera01', 0),
 ('mushipa01', 0),
 ('nerishe01', 0),
 ('odorija01', 0),
 ('pareden01', 0),
 ('perezjo01', 0),
 ('pressry01', 0),
 ('smithwi04', 0),
 ('stanery01', 0),
 ('taylobl01', 0),
 ('urquijo01', 0),
 ('valdefr01', 0),
 ('verlaju01', 0)]

playerID	RBI
tuckeky01	107
alvaryo01	97
bregmal01	93
penaje02	63
altuvjo01	57
gourryu01	53
maldoma01	45
mccorch01	44
diazal02	38
brantmi02	26
mancitr01	22
dubonma01	16
meyerja02	15
sirijo01	10
vazquch01	10
henslda01	5
matijjj01	5
leeko01	4
castrja01	3
diazya02	1
goodrni01	1
abreubr01	0
baezpe01	0
bielabr01	0
blancro01	0
brownhu01	0
garcilu05	0
javiecr01	0
jonesta01	0
martise01	0
matonph01	0
mcculla02	0
montera01	0
mushipa01	0
nerishe01	0
odorija01	0
pareden01	0
perezjo01	0
pressry01	0
smithwi04	0
stanery01	0
taylobl01	0
urquijo01	0
valdefr01	0
verlaju01	0


C.

In [26]:
%%sql
SELECT b.playerID, b.RBI, p.nameFirst, p.nameLast FROM batting AS b JOIN people AS p ON b.playerID = p.playerID WHERE teamID = 'HOU' AND yearID = 2022 ORDER BY b.RBI DESC LIMIT 1;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('tuckeky01', 107, 'Kyle', 'Tucker')]

playerID	RBI	nameFirst	nameLast
tuckeky01	107	Kyle	Tucker

## Question 3

A.

In [10]:
%%sql
SELECT * FROM allstarfull WHERE playerID = 'tuckeky01';

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('tuckeky01', 2022, 0, 'NLS202207190', 'HOU', 'AL', 1, '')]

playerID	yearID	gameNum	gameID	teamID	lgID	GP	startingPos
tuckeky01	2022	0	NLS202207190	HOU	AL	1	


B.

In [14]:
%%sql
SELECT playerID, COUNT(*) as all_star_appearances FROM allstarfull GROUP BY playerID ORDER BY all_star_appearances DESC LIMIT 1;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('mayswi01', 26)]

playerID	all_star_appearances
mayswi01	26


## Question 4

In [18]:
%%sql
SELECT playerID, AVG(salary) AS average_salary FROM salaries GROUP BY playerID ORDER BY average_salary DESC LIMIT 100;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('tanakma01', 22000000.0),
 ('rodrial01', 18109829.636363637),
 ('howarry01', 15525500.0),
 ('teixema01', 15305357.142857144),
 ('sabatcc01', 14576190.4),
 ('jeterde01', 13927268.05263158),
 ('fieldpr01', 13901318.181818182),
 ('cabremi01', 13457901.642857144),
 ('mauerjo01', 13232692.307692308),
 ('cespeyo01', 13165609.2),
 ('verlaju01', 12974090.909090908),
 ('pujolal01', 12752527.25),
 ('kurodhi01', 12580817.57142857),
 ('hamelco01', 12540000.0),
 ('hernafe02', 12529844.181818182),
 ('linceti01', 12444375.0),
 ('hollima01', 12040131.083333334),
 ('greinza01', 12025169.166666666),
 ('canoro01', 11806527.272727273),
 ('priceda01', 11693042.285714285),
 ('zambrca01', 11495000.0),
 ('beltrca01', 11432376.777777778),
 ('soriaal01', 11425714.285714285),
 ('kershcl01', 11301875.0),
 ('hamiljo03', 10969057.9),
 ('ramirma02', 10885672.05263158),
 ('santajo01', 10803523.333333334),
 ('crawfca02', 10699530.5),
 ('kempma01', 10660600.0),
 ('hallaro01', 10642261.857142856),
 ('zitoba01', 10541153.846153846),
 ('cainma01', 10431333.181818182),
 ('wrighda03', 10354447.666666666),
 ('ramirha01', 10333454.545454545),
 ('gonzaad01', 10324148.75),
 ('suzukic01', 10258217.6875),
 ('hunteto01', 10091470.588235294),
 ('pencehu01', 9848333.333333334),
 ('poseybu01', 9827963.0),
 ('delgaca01', 9753266.666666666),
 ('weaveje02', 9751500.0),
 ('fukudko01', 9700000.0),
 ('beltrad01', 9638947.368421054),
 ('heltoto01', 9637647.05882353),
 ('ethiean01', 9511200.0),
 ('utleych01', 9472692.153846154),
 ('vottojo01', 9453829.444444444),
 ('ellsbja01', 9358952.333333334),
 ('leeca01', 9341071.42857143),
 ('matsuhi01', 9250000.0),
 ('peavyja01', 9200876.5),
 ('oswalro01', 9195000.0),
 ('beckejo02', 9151201.846153846),
 ('martipe02', 9141224.0625),
 ('abreujo02', 9110889.0),
 ('wellsve01', 9016835.153846154),
 ('lackejo01', 8996011.857142856),
 ('riverma01', 8914834.210526315),
 ('ordonma01', 8898049.733333332),
 ('ortizda01', 8889583.333333334),
 ('berkmla01', 8857500.0),
 ('lestejo01', 8855550.0),
 ('reyesjo01', 8830076.923076924),
 ('leecl02', 8821275.0),
 ('morneju01', 8816363.636363637),
 ('buehrma01', 8784285.714285715),
 ('zimmery01', 8759200.0),
 ('darviyu01', 8750000.0),
 ('ramirar01', 8699117.647058824),
 ('dunnad01', 8668846.153846154),
 ('wainwad01', 8657909.090909092),
 ('scherma01', 8604357.125),
 ('bondsba01', 8556605.545454545),
 ('bagweje01', 8542267.933333334),
 ('tulowtr01', 8538100.0),
 ('burneaj01', 8486558.823529411),
 ('jonesch06', 8477606.65),
 ('gonzaca01', 8445754.857142856),
 ('mccanbr01', 8394560.545454545),
 ('werthja01', 8378064.285714285),
 ('guerrvl01', 8369430.333333333),
 ('johnsra05', 8359524.714285715),
 ('zimmejo02', 8344166.666666667),
 ('drewjd01', 8314745.230769231),
 ('uptonju01', 8195925.777777778),
 ('papeljo01', 8188412.0),
 ('wilsocj01', 8144500.2727272725),
 ('grandcu01', 8090454.545454546),
 ('pettian01', 8078377.411764706),
 ('martivi01', 8071276.923076923),
 ('mussimi01', 8029645.5),
 ('piazzmi01', 8011733.466666667),
 ('sheffga01', 8000407.142857143),
 ('markani01', 7902909.090909091),
 ('castiru01', 7847476.333333333),
 ('kawakke01', 7833666.0),
 ('riosal01', 7832272.7272727275),
 ('posadjo01', 7830566.666666667),
 ('phillbr01', 7806824.181818182),
 ('youngmi02', 7789116.416666667)]

playerID	average_salary
tanakma01	22000000.0
rodrial01	18109829.636363637
howarry01	15525500.0
teixema01	15305357.142857144
sabatcc01	14576190.4
jeterde01	13927268.05263158
fieldpr01	13901318.181818182
cabremi01	13457901.642857144
mauerjo01	13232692.307692308
cespeyo01	13165609.2
verlaju01	12974090.909090908
pujolal01	12752527.25
kurodhi01	12580817.57142857
hamelco01	12540000.0
hernafe02	12529844.181818182
linceti01	12444375.0
hollima01	12040131.083333334
greinza01	12025169.166666666
canoro01	11806527.272727273
priceda01	11693042.285714285
zambrca01	11495000.0
beltrca01	11432376.777777778
soriaal01	11425714.285714285
kershcl01	11301875.0
hamiljo03	10969057.9
ramirma02	10885672.05263158
santajo01	10803523.333333334
crawfca02	10699530.5
kempma01	10660600.0
hallaro01	10642261.857142856
zitoba01	10541153.846153846
cainma01	10431333.181818182
wrighda03	10354447.666666666
ramirha01	10333454.545454545
gonzaad01	10324148.75
suzukic01	10258217.6875
hunteto01	10091470.588235294
pencehu01	9848333.333333334
poseybu01	9827963.0
delgaca01	9753266.666666666
weaveje02	9751500.0
fukudko01	9700000.0
beltrad01	9638947.368421054
heltoto01	9637647.05882353
ethiean01	9511200.0
utleych01	9472692.153846154
vottojo01	9453829.444444444
ellsbja01	9358952.333333334
leeca01	9341071.42857143
matsuhi01	9250000.0
peavyja01	9200876.5
oswalro01	9195000.0
beckejo02	9151201.846153846
martipe02	9141224.0625
abreujo02	9110889.0
wellsve01	9016835.153846154
lackejo01	8996011.857142856
riverma01	8914834.210526315
ordonma01	8898049.733333332
ortizda01	8889583.333333334
berkmla01	8857500.0
lestejo01	8855550.0
reyesjo01	8830076.923076924
leecl02	8821275.0
morneju01	8816363.636363637
buehrma01	8784285.714285715
zimmery01	8759200.0
darviyu01	8750000.0
ramirar01	8699117.647058824
dunnad01	8668846.153846154
wainwad01	8657909.090909092
scherma01	8604357.125
bondsba01	8556605.545454545
bagweje01	8542267.933333334
tulowtr01	8538100.0
burneaj01	8486558.823529411
jonesch06	8477606.65
gonzaca01	8445754.857142856
mccanbr01	8394560.545454545
werthja01	8378064.285714285
guerrvl01	8369430.333333333
johnsra05	8359524.714285715
zimmejo02	8344166.666666667
drewjd01	8314745.230769231
uptonju01	8195925.777777778
papeljo01	8188412.0
wilsocj01	8144500.2727272725
grandcu01	8090454.545454546
pettian01	8078377.411764706
martivi01	8071276.923076923
mussimi01	8029645.5
piazzmi01	8011733.466666667
sheffga01	8000407.142857143
markani01	7902909.090909091
castiru01	7847476.333333333
kawakke01	7833666.0
riosal01	7832272.7272727275
posadjo01	7830566.666666667
phillbr01	7806824.181818182
youngmi02	7789116.416666667

## Question 5

10 Highest paid athletes

In [29]:
%%sql
SELECT playerID, MAX(salary) AS max_salary FROM salaries GROUP BY playerID ORDER BY max_salary DESC LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('rodrial01', 33000000.0),
 ('kershcl01', 33000000.0),
 ('greinza01', 31799030.0),
 ('priceda01', 30000000.0),
 ('verlaju01', 28000000.0),
 ('cabremi01', 28000000.0),
 ('cespeyo01', 27328046.0),
 ('wellsve01', 26187500.0),
 ('hernafe02', 25857143.0),
 ('sabatcc01', 25000000.0)]

playerID	max_salary
rodrial01	33000000.0
kershcl01	33000000.0
greinza01	31799030.0
priceda01	30000000.0
verlaju01	28000000.0
cabremi01	28000000.0
cespeyo01	27328046.0
wellsve01	26187500.0
hernafe02	25857143.0
sabatcc01	25000000.0

# Project 10

## Question 1

In [2]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [3]:
%%sql
SELECT yearID AS year, COUNT(*) AS num_of_players FROM appearances WHERE (yearID <= 2022) AND (yearID >= 2018) GROUP BY yearID ORDER BY num_of_players DESC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(2021, 1706), (2022, 1683), (2019, 1567), (2018, 1533), (2020, 1360)]

year	num_of_players
2021	1706
2022	1683
2019	1567
2018	1533
2020	1360


## Question 2

A.

In [4]:
%%sql
SELECT AVG(t.attendance) as average_attendance, t.teamID FROM teams t JOIN
(SELECT AVG(attendance) as totalavgattend FROM teams WHERE yearID = 2022) AS s
WHERE t.yearID = 2022 GROUP BY t.teamID 
HAVING (average_attendance >= s.totalavgattend)
ORDER BY average_attendance DESC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(3861408.0, 'LAN'),
 (3320551.0, 'SLN'),
 (3136207.0, 'NYA'),
 (3129931.0, 'ATL'),
 (2987470.0, 'SDN'),
 (2688998.0, 'HOU'),
 (2653830.0, 'TOR'),
 (2625089.0, 'BOS'),
 (2616780.0, 'CHN'),
 (2597428.0, 'COL'),
 (2564737.0, 'NYN'),
 (2482686.0, 'SFN'),
 (2457461.0, 'LAA'),
 (2422420.0, 'MIL'),
 (2287267.0, 'SEA'),
 (2276736.0, 'PHI')]

average_attendance	teamID
3861408.0	LAN
3320551.0	SLN
3136207.0	NYA
3129931.0	ATL
2987470.0	SDN
2688998.0	HOU
2653830.0	TOR
2625089.0	BOS
2616780.0	CHN
2597428.0	COL
2564737.0	NYN
2482686.0	SFN
2457461.0	LAA
2422420.0	MIL
2287267.0	SEA
2276736.0	PHI


B.

In [5]:
%%sql
SELECT AVG(attendance) as totalavgattend FROM teams WHERE yearID = 2022;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(2151888.6,)]

totalavgattend
2151888.6


## Question 3

In [6]:
%%sql
SELECT teamID, yearID, CAST(W AS FLOAT) / (W + L) AS winning_per FROM teams WHERE teamID = 'LAN' AND yearID = 2022 GROUP BY teamID;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('LAN', 2022, 0.6851851851851852)]

teamID	yearID	winning_per
LAN	2022	0.6851851851851852


## Question 4

In [7]:
%%sql
SELECT teamID, SUM(W) * 1.0 / SUM(W + L) AS winning_per FROM teams WHERE (yearID <= 2022) AND (yearID >= 2012) GROUP BY teamID HAVING SUM(W) * 1.0 / SUM(W + L) > 0.55 ORDER BY winning_per DESC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('LAN', 0.6049970255800119),
 ('NYA', 0.5672619047619047),
 ('SLN', 0.5578069129916567)]

teamID	winning_per
LAN	0.6049970255800119
NYA	0.5672619047619047
SLN	0.5578069129916567


## Question 5

A.

In [15]:
%%sql
SELECT playerID, COUNT(*) AS number_of_all_star_games FROM allstarfull GROUP BY playerID HAVING COUNT(*) >= 20 ORDER BY number_of_all_star_games DESC;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('mayswi01', 26), ('musiast01', 24), ('aaronha01', 24)]

playerID	number_of_all_star_games
mayswi01	26
musiast01	24
aaronha01	24

B.

In [23]:
%%sql
SELECT A.playerID, P.total_awards, P.first_year, P.last_year FROM 
(SELECT playerID, COUNT(*) AS total_awards, MIN(yearID) AS first_year, MAX(yearID) AS last_year FROM awardsplayers GROUP BY playerID) 
AS P INNER JOIN awardsplayers AS A ON P.playerID = A.playerID
GROUP BY A.playerID ORDER BY P.total_awards DESC, P.first_year ASC 
LIMIT 5;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('bondsba01', 47, 1990, 2004),
 ('willite01', 39, 1939, 1958),
 ('dimagjo01', 36, 1936, 1950),
 ('musiast01', 34, 1943, 1958),
 ('johnswa01', 33, 1909, 1925)]

playerID	total_awards	first_year	last_year
bondsba01	47	1990	2004
bondsba01	47	1990	2004
bondsba01	47	1990	2004
bondsba01	47	1990	2004
bondsba01	47	1990	2004


# Project 11

## Question 1

In [1]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [19]:
%%sql
SELECT p.playerID, p.nameFirst, p.nameLast, CAST(b.H AS FLOAT)/CAST(b.AB AS FLOAT) AS battingAverage
FROM batting b
INNER JOIN people p ON b.playerID = p.playerID
WHERE b.yearID = 2022 AND b.AB > 0
ORDER BY battingAverage ASC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('alvarjo03', 'Jose', 'Alvarado', 0.0),
 ('anderty01', 'Tyler', 'Anderson', 0.0),
 ('arauzjo01', 'Jonathan', 'Arauz', 0.0),
 ('blanktr01', 'Travis', 'Blankenhorn', 0.0),
 ('changyu01', 'Yu', 'Chang', 0.0),
 ('dawsoro01', 'Ronnie', 'Dawson', 0.0),
 ('diazyu01', 'Yusniel', 'Diaz', 0.0),
 ('difowi01', 'Wilmer', 'Difo', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0)]

playerID	nameFirst	nameLast	battingAverage
alvarjo03	Jose	Alvarado	0.0
anderty01	Tyler	Anderson	0.0
arauzjo01	Jonathan	Arauz	0.0
blanktr01	Travis	Blankenhorn	0.0
changyu01	Yu	Chang	0.0
dawsoro01	Ronnie	Dawson	0.0
diazyu01	Yusniel	Diaz	0.0
difowi01	Wilmer	Difo	0.0
faircst01	Stuart	Fairchild	0.0
faircst01	Stuart	Fairchild	0.0


## Question 2

A.

In [40]:
%%sql
SELECT playerID, G_p FROM appearances WHERE yearID = 2022 AND G_p > 0;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('abbotco01', 16),
 ('abreual01', 4),
 ('abreual01', 22),
 ('abreual01', 7),
 ('abreubr01', 55),
 ('acevedo01', 70),
 ('adamja01', 67),
 ('adamsau02', 2),
 ('adonjo01', 14),
 ('adriaeh01', 1),
 ('akinke01', 45),
 ('alberha01', 10),
 ('alcaljo01', 2),
 ('alcansa01', 32),
 ('alexaja01', 18),
 ('alexasc01', 17),
 ('alexaty01', 27),
 ('alexyaj01', 4),
 ('allarko01', 10),
 ('alldrca01', 1),
 ('allenlo01', 3),
 ('allenlo01', 4),
 ('almonye01', 33),
 ('alvarjo02', 21),
 ('alvarjo03', 59),
 ('alvarrj01', 1),
 ('alzolad01', 6),
 ('anderch01', 9),
 ('anderia01', 22),
 ('andersh01', 1),
 ('anderty01', 30),
 ('appelma01', 6),
 ('aranovi01', 43),
 ('archech01', 25),
 ('arciaor01', 1),
 ('arihako01', 5),
 ('armstsh01', 7),
 ('armstsh01', 43),
 ('ashbyaa01', 27),
 ('ashcrgr01', 19),
 ('assadja01', 9),
 ('astudwi01', 2),
 ('avilape01', 2),
 ('baezmi01', 2),
 ('baezpe01', 3),
 ('bakerbr01', 66),
 ('bandaan01', 2),
 ('bandaan01', 23),
 ('bandaan01', 7),
 ('banksta01', 35),
 ('banuema01', 4),
 ('banuema01', 31),
 ('bardda01', 57),
 ('bardlu01', 1),
 ('bardlu01', 8),
 ('barkelu01', 3),
 ('barlojo01', 35),
 ('barlosc01', 69),
 ('barneja01', 22),
 ('barneja01', 1),
 ('barnema01', 44),
 ('barnhtu01', 1),
 ('barraky01', 8),
 ('barrija01', 35),
 ('bassan01', 45),
 ('bassan01', 28),
 ('bassich01', 30),
 ('battema01', 2),
 ('baumami01', 13),
 ('bautife01', 65),
 ('bazared01', 12),
 ('bazsh01', 6),
 ('beaslje01', 9),
 ('bednada01', 45),
 ('beedety01', 25),
 ('beedety01', 6),
 ('beeksja01', 42),
 ('bellaan01', 59),
 ('bellobr01', 13),
 ('bendean01', 22),
 ('bernabr01', 2),
 ('berrijo01', 32),
 ('bethach01', 1),
 ('bethach01', 3),
 ('bickfph01', 60),
 ('biebesh01', 31),
 ('bielabr01', 5),
 ('birdja01', 38),
 ('blachty01', 24),
 ('blackpa01', 21),
 ('blancro01', 7),
 ('bleieri01', 55),
 ('bolanro01', 8),
 ('borucry01', 21),
 ('borucry01', 11),
 ('boxbebr01', 70),
 ('boydma01', 10),
 ('brachsi01', 3),
 ('bradiky01', 23),
 ('bradlar01', 21),
 ('bradlja02', 1),
 ('brashma01', 39),
 ('brasiry01', 68),
 ('braulst01', 9),
 ('brazohu01', 27),
 ('brebbjo01', 76),
 ('brentja01', 8),
 ('briceau01', 4),
 ('briesbe01', 15),
 ('brighje01', 16),
 ('brittza01', 3),
 ('brogdco01', 47),
 ('brookaa01', 5),
 ('brossmi01', 3),
 ('brownhu01', 7),
 ('brubajt01', 28),
 ('bruihju01', 24),
 ('bubickr01', 28),
 ('buehlwa01', 12),
 ('bumgama01', 30),
 ('bummeaa01', 32),
 ('bundydy01', 29),
 ('burkebr01', 52),
 ('burneco01', 33),
 ('burrry01', 8),
 ('bushma01', 25),
 ('bushma01', 40),
 ('buttojo01', 1),
 ('cabreed02', 14),
 ('cabrege01', 39),
 ('canoye01', 3),
 ('canoye01', 10),
 ('caratvi01', 1),
 ('carltdr01', 5),
 ('carraca01', 29),
 ('castada01', 10),
 ('castehu01', 11),
 ('castery01', 3),
 ('castidi01', 59),
 ('castidi02', 2),
 ('castijo03', 1),
 ('castilu02', 14),
 ('castilu02', 11),
 ('castilu03', 3),
 ('castima03', 5),
 ('castima03', 9),
 ('castran02', 12),
 ('castrha01', 4),
 ('castrke01', 8),
 ('castrke01', 2),
 ('castrmi01', 34),
 ('cavalca01', 1),
 ('ceasedy01', 32),
 ('cessalu01', 46),
 ('chacijh01', 35),
 ('chafian01', 64),
 ('changyu01', 2),
 ('chapmar01', 43),
 ('chargjt01', 21),
 ('chaveje01', 46),
 ('chaveje01', 3),
 ('chaveje01', 11),
 ('chiriyo01', 2),
 ('cimbead01', 77),
 ('cishest01', 69),
 ('cisnejo01', 28),
 ('civalaa01', 20),
 ('clarkta01', 47),
 ('claseem01', 77),
 ('claudal01', 3),
 ('claysa01', 1),
 ('claysa01', 6),
 ('cleavga01', 4),
 ('cleavga01', 13),
 ('clemeer01', 2),
 ('clemeko01', 7),
 ('clevimi01', 23),
 ('clippty01', 4),
 ('cobbal01', 28),
 ('colege01', 33),
 ('colemdy01', 68),
 ('colomal01', 53),
 ('contrro01', 21),
 ('coonrsa01', 12),
 ('corbipa01', 31),
 ('cortene01', 28),
 ('cottojh01', 25),
 ('cottojh01', 5),
 ('couloda01', 10),
 ('cousija01', 12),
 ('crawfku01', 21),
 ('crickky01', 14),
 ('crismna01', 50),
 ('criswco01', 1),
 ('crowewi01', 60),
 ('cruzfe01', 14),
 ('cruzje01', 7),
 ('cuasjo01', 47),
 ('cuetojo01', 25),
 ('culbech01', 2),
 ('curryxz01', 2),
 ('cyrty01', 11),
 ('cyrty01', 1),
 ('danisty01', 32),
 ('darviyu01', 30),
 ('davidtu0

B.


In [35]:
%%sql
SELECT p.playerID, p.nameFirst, p.nameLast, CAST(b.H AS FLOAT)/CAST(b.AB AS FLOAT) AS battingAverage
FROM batting b
INNER JOIN people p ON b.playerID = p.playerID
WHERE b.yearID = 2022 
AND b.AB > 0 
AND b.playerID NOT IN (SELECT playerID FROM appearances WHERE yearID = 2022 AND G_p > 0)
ORDER BY battingAverage ASC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('arauzjo01', 'Jonathan', 'Arauz', 0.0),
 ('blanktr01', 'Travis', 'Blankenhorn', 0.0),
 ('dawsoro01', 'Ronnie', 'Dawson', 0.0),
 ('diazyu01', 'Yusniel', 'Diaz', 0.0),
 ('difowi01', 'Wilmer', 'Difo', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0),
 ('fermifr01', 'Freddy', 'Fermin', 0.0),
 ('godoyjo01', 'Jose', 'Godoy', 0.0),
 ('guzmaro01', 'Ronald', 'Guzman', 0.0)]

playerID	nameFirst	nameLast	battingAverage
arauzjo01	Jonathan	Arauz	0.0
blanktr01	Travis	Blankenhorn	0.0
dawsoro01	Ronnie	Dawson	0.0
diazyu01	Yusniel	Diaz	0.0
difowi01	Wilmer	Difo	0.0
faircst01	Stuart	Fairchild	0.0
faircst01	Stuart	Fairchild	0.0
fermifr01	Freddy	Fermin	0.0
godoyjo01	Jose	Godoy	0.0
guzmaro01	Ronald	Guzman	0.0

## Question 3

In [37]:
%%sql
SELECT p.playerID, p.nameFirst, p.nameLast, CAST(b.H AS FLOAT)/CAST(b.AB AS FLOAT) AS battingAverage
FROM batting b
INNER JOIN people p ON b.playerID = p.playerID
LEFT JOIN appearances a ON b.playerID = a.playerID AND b.yearID = a.yearID AND a.G_p > 0
WHERE b.yearID = 2022 AND b.AB > 0 AND a.playerID IS NULL
ORDER BY battingAverage ASC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('arauzjo01', 'Jonathan', 'Arauz', 0.0),
 ('blanktr01', 'Travis', 'Blankenhorn', 0.0),
 ('dawsoro01', 'Ronnie', 'Dawson', 0.0),
 ('diazyu01', 'Yusniel', 'Diaz', 0.0),
 ('difowi01', 'Wilmer', 'Difo', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0),
 ('faircst01', 'Stuart', 'Fairchild', 0.0),
 ('fermifr01', 'Freddy', 'Fermin', 0.0),
 ('godoyjo01', 'Jose', 'Godoy', 0.0),
 ('guzmaro01', 'Ronald', 'Guzman', 0.0)]

playerID	nameFirst	nameLast	battingAverage
arauzjo01	Jonathan	Arauz	0.0
blanktr01	Travis	Blankenhorn	0.0
dawsoro01	Ronnie	Dawson	0.0
diazyu01	Yusniel	Diaz	0.0
difowi01	Wilmer	Difo	0.0
faircst01	Stuart	Fairchild	0.0
faircst01	Stuart	Fairchild	0.0
fermifr01	Freddy	Fermin	0.0
godoyjo01	Jose	Godoy	0.0
guzmaro01	Ronald	Guzman	0.0


## Question 4

In [39]:
%%sql
SELECT AVG(CAST(b.H AS FLOAT)/CAST(b.AB AS FLOAT)) AS averageBattingAverage
FROM batting b
LEFT JOIN appearances a ON b.playerID = a.playerID AND b.yearID = a.yearID AND a.G_p > 0
WHERE b.yearID = 2022 AND b.AB > 0 AND a.playerID IS NULL;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(0.2128322341979076,)]

averageBattingAverage
0.2128322341979076